### PR TITLE
df.groupby().agg() feature

### LINKED ISSUE #15967: [FEA] Support named aggregations in `df.groupby().agg()`
**Is your feature request related to a problem? Please describe.**
I wish cudf python supported column name keywords in `df.groupby().agg()`

[See pandas docs here](https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.aggregate.html)
> Here the output has one column for each element in **kwargs. The name of the column is keyword, whereas the value determines the aggregation used to compute the values in the column.

**Describe the solution you'd like**
Let's add support for column name keywords in `df.groupby().agg()`. This should be a python-only change for API convenience. It will prevent host fallback in `cudf.pandas`

**Describe alternatives you've considered**
Instead of:
`df.groupby('a').agg(mycol=('b','sum'))`
we can say
`df.groupby('a').agg({'b': 'sum'}).rename(columns={'b': 'named_agg'})`
and renaming columns doesn't work with a list of aggregations on the same column
`df.groupby('a').agg({'b': ['sum', 'mean']}).rename(columns={???: 'sum'})`

It seems like we could write a convenience API here to capture any `kwargs` and their tuples to make new columns.

**Additional context**
Many nice pandas queries for TPC-H were published in the [fireducks-dev](https://github.com/fireducks-dev/polars-tpch/tree/fireducks/queries/pandas) repo, and the `agg` column name keywords style is common.

### PR DESCRIPTION
**Pull request by NeilGeorge1**
### Added **kwargs Support in df.groupby().agg()
Implemented support for **kwargs in df.groupby().agg() to provide greater flexibility and customization options during aggregation operations. 

This issue closes #15967

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by vyasr** [comment]
Thanks for the contribution @NeilGeorge1! Could you add some tests of some different cases? I know you won't be able to run them on your machine without a GPU, but you should be able to write them using pandas code locally and then simply switch to using cudf at the end before commiting. Then we can use your tests to validate the behavior here. I would recommend trying a few different more complex cases such as including multiple output columns.

**Comment by NeilGeorge1** [comment]
> Thanks for the contribution @NeilGeorge1! Could you add some tests of some different cases? I know you won't be able to run them on your machine without a GPU, but you should be able to write them using pandas code locally and then simply switch to using cudf at the end before commiting. Then we can use your tests to validate the behavior here. I would recommend trying a few different more complex cases such as including multiple output columns.
> 
> Also, right now you've made the change to the dask object, but we want the functionality in cudf itself. Have a look at [this file](https://github.com/rapidsai/cudf/blob/branch-24.08/python/cudf/cudf/core/groupby/groupby.py), which is where I think you'd want to add the new code.

Yeah sure @vyasr. I don't have any prior experience writing test cases before but I shall try my best.
Also should I just leave the code in dask as it is or should i remove it?

**Comment by vyasr** [comment]
I think @galipremsagar or @wence- can probably give you a better answer about how best we want this to look in dask.

**Comment by wence-** [comment]
> I think @galipremsagar or @wence- can probably give you a better answer about how best we want this to look in dask.

Let's first get the cudf implementation right first, it might then be we don't have to do anything dask-dataframe.

To do this, I think we want to mimic the way the pandas API behaves (https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.agg.html)

I don't think we support arguments and engine to `agg` in cudf, so we want to support this signature:

```
def agg(self, func, **kwargs):
```

The behaviour of pandas is:

- If `func` is not `None`, then `kwargs` are ignored (silently)
- If `func` is `None`, then `kwargs` are used with the key defining the output column name, and the value defining the `(column, aggregation)` pair.

This preprocessing should happen in `cudf/python/cudf/cudf/core/groupby/groupby.py` in the `agg` method (around line 600).

I think the way to attack this is to let `_normalize_aggs` take both `func` and `kwargs`, and produce appropriately preprocessed results. There's a little bit of a dance we need to do to get the right names out, but it's certainly possible.

**Comment by NeilGeorge1** [comment]
Understood @wence- 

I will start by reading the documentation to grasp the required changes and proceed to implement them. 

Following this, I will add few more test cases taking reference from the examples provided in the groupby.py file into python/cudf/cudf/tests/groupby/test_agg.py as requested by @vyasr . 

Given that my PC lacks a GPU, I will then proceed by submitting a pull request for the changes. 

Is there anything else I need to consider?

**Comment by NeilGeorge1** [comment]
Understood @wence- 

I will start by reading the documentation to grasp the required changes and proceed to implement them.

Following this, I will add few more test cases taking reference from the examples provided in the groupby.py file into `python/cudf/cudf/tests/groupby/test_agg.py` as requested by @vyasr .

Given that my PC lacks a GPU, I will then run these tests using only `pandas` and then when all test cases pass, switch to `cudf` before committing and making a pull request.

Is there anything else I need to consider?

**Comment by wence-** [comment]
> Understood @wence-
> 
> I will start by reading the documentation to grasp the required changes and proceed to implement them.
> 
> Following this, I will add few more test cases taking reference from the examples provided in the `groupby.py` file into `python/cudf/cudf/tests/groupby/test_agg.py` as requested by @vyasr .
> 
> Given that my PC lacks a GPU, I will then run these tests using only `pandas` and then when all test cases pass, switch to `cudf` before committing and making a pull request.
> 
> Is there anything else I need to consider?

That sounds about right. If you have queries about the (admittedly underdocumented) internals of the groupby datastructures, please ask here.

**Comment by NeilGeorge1** [comment]
> > Understood @wence-
> > I will start by reading the documentation to grasp the required changes and proceed to implement them.
> > Following this, I will add few more test cases taking reference from the examples provided in the `groupby.py` file into `python/cudf/cudf/tests/groupby/test_agg.py` as requested by @vyasr .
> > Given that my PC lacks a GPU, I will then run these tests using only `pandas` and then when all test cases pass, switch to `cudf` before committing and making a pull request.
> > Is there anything else I need to consider?
> 
> That sounds about right. If you have queries about the (admittedly underdocumented) internals of the groupby datastructures, please ask here.

Sure thing! Will do thanks!

File: python/dask_cudf/dask_cudf/groupby.py [hunk 1]
```diff
@@ -196,8 +196,11 @@ def last(self, split_every=None, split_out=1):
     @_deprecate_shuffle_kwarg
     @_dask_cudf_nvtx_annotate
     def aggregate(
-        self, arg, split_every=None, split_out=1, shuffle_method=None
+        self, arg, split_every=None, split_out=1, shuffle_method=None, **kwargs
     ):
+        if kwargs:
+            arg = {col_name: agg_func for col_name, (col, agg_func) in kwargs.items()}
+
```
Comment (rjzamora) [inline python/dask_cudf/dask_cudf/groupby.py:203]: I know you are already revising this PR, but just a quick note: These changes are being made to the legacy dask-cudf API (which will soon be deprecated and then completely removed). The new (default) expression-based API is defined [in expr/_groupby.py](https://github.com/rapidsai/cudf/blob/fa8284ddb2de808573d5b21cc9e650578ddf6acc/python/dask_cudf/dask_cudf/expr/_groupby.py#L92).

Comment (NeilGeorge1) [reply python/dask_cudf/dask_cudf/groupby.py:203]: Ok Sure working on cudf now

**Comment by vyasr** [comment]
@NeilGeorge1 have you had any luck here? Do you need any pointers?

**Comment by Matt711** [comment]
I'll open a new PR for this.

**Pull request closed without merging**